src/crafting.h

```diff
@@ -104,6 +104,55 @@ struct recipe {
 
 };
 
+enum usage {
+    use_from_map = 1,
+    use_from_player = 2,
+    use_from_both = 1 | 2,
+    use_from_none = 4,
+    cancel = 5
+};
+
+template<typename CompType = component>
+struct comp_selection {
+    usage use_from = use_from_none;
+    CompType comp;
+
+    std::string nname();
+};
+using item_selection = comp_selection<item_comp>;
+using tool_selection = comp_selection<tool_comp>;
+
+class craft_command {
+    public:
+        const recipe* rec = nullptr;
+        int batch_size = 0;
+        bool is_long = false;
+        player* crafter; // This is mainly here for maintainability reasons.
+
+        craft_command() {}
+        craft_command( const recipe* to_make, int batch_size, bool is_long, player* crafter ) :
+            rec( to_make ), batch_size( batch_size ), is_long( is_long ), crafter( crafter ) {}
+
+        void execute();
+        std::list<item> consume_components();
+
+        bool has_cached_selections()
+        {
+            return !item_selections.empty() || !tool_selections.empty();
+        }
+
+        bool empty()
```

Comment:
It might not matter now, but this function (and `has_cached_selections` above and `comp_selection::nname`) should be const. The check functions below maybe as well, but I haven't looked into them.

It looks strange to see as `has` or `is` or `check` function being non-const, the verb implies only read-access to the data.